HPCC-20914 Log why extra log line is added to ESP log

wangkx opened the pull request:
When the logLevel is set to 1 and the logRequests is set to
false, ESP will log requests if there is some exception or
the processing time is longer than a pre-defined value. In
this fix, ESP will also log why requests are logged.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-20914 Log why extra log line is added to ESP log

When the logLevel is set to 1 and the logRequests is set to
false, ESP will log requests if there is some exception or
the processing time is longer than a pre-defined value. In
this fix, ESP will also log why requests are logged.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@afishbeck please review.

HPCCSmoketest commented:
Process of PR-12611, label: h20914 starts now.
The reason of this test is: Did not build yet.
Commit ID: 77452dd319b7aef896c38c0381ad1ea8025f621d
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 77452dd319b7aef896c38c0381ad1ea8025f621d
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |835 |835 |0 |176 sec  (00:02:56) |
|test (thor) |759 |759 |0 |623 sec  (00:10:23) |
|test (roxie) |911 |911 |0 |232 sec  (00:03:52) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  212 sec (00:03:32) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1343 sec (00:22:23) |   20 sec (00:00:20) | 1629 sec (00:27:09) |

afishbeck commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -73,7 +73,12 @@ CEspHttpServer::~CEspHttpServer()
 
-                DBGLOG("Request[%s]", logStr.str());
+                if (ctx->queryHasException())
+                    DBGLOG("Log request due to exception. Request[%s]", logStr.str());
```
Perhaps simplify slightly to "Request with exception: Request[%s]" and "Request with long processing time: %u seconds. Request[%s]"?

Unless you think its important to be able to search or parse for "Log Request"?

wangkx replied:
@afishbeck do you mean "Request with exception. Request[%s]"?

afishbeck reviewed (commented):
@wangkx very minor comment.  Otherwise change is fine.

richardkchapman commented:
I'm a bit concerned that we may end up logging requests (which could contain sensitive information that should not be logged) when logging of requests has been explicitly turned off (perhaps because they contain sensitive information...). 

I would have thought we need a way to say loggingRequests=never|only-ones-with-issues|all

wangkx committed:
HPCC-20914 Revise based on review

1. Change the log messages;
2. Change logRequest settings from true/false to never|
  only-ones-with-issues|all and add ESP code to handle
  that;
3. Only log ESP request/response based on logRequest
  logResponse settings.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-12611, label: h20914 starts now.
The reason of this test is: New commit.
Commit ID: eb5cf6e41dc43ae17d899b7d4dd9088c2fb7774c
Estimated completion time is ~0.5 hour(s)

wangkx commented:
@afishbeck please review my changes.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: eb5cf6e41dc43ae17d899b7d4dd9088c2fb7774c
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |33 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |47 sec  (00:00:47) |
|setup (roxie) |11 |11 |0 |30 sec  (00:00:30) |
|test (hthor) |835 |835 |0 |181 sec  (00:03:01) |
|test (thor) |759 |759 |0 |624 sec  (00:10:24) |
|test (roxie) |911 |911 |0 |235 sec  (00:03:55) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   36 sec (00:00:36) |  211 sec (00:03:31) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1357 sec (00:22:37) |   19 sec (00:00:19) | 1643 sec (00:27:23) |

afishbeck commented on esp/platform/espcontext.cpp:
```diff
@@ -820,6 +820,22 @@ static IEspContainer*& getContainer()
+        return LogRequestsNever;
+    if (strieq(req, "only-ones-with-issues"))
+        return LogRequestsWithIssuesOnly;
+    if (strieq(req, "true"))
```
should probably support "1" as a boolean true value.

richardkchapman replied:
There's a standard function we use for such situations (I can't remember its name)

wangkx replied:
I guess I can use getEnum().

afishbeck commented on esp/platform/espcontext.cpp:
```diff
@@ -820,6 +820,22 @@ static IEspContainer*& getContainer()
 
+LogRequest readLogRequest(char const* req)
+{
+    if (isEmptyString(req))
```
should probably treat old values of logRequests=false and logRequests=0 as LogRequestsNever.

afishbeck reviewed (changes requested):
@wangkx a couple comments.

wangkx committed:
HPCC-20914 Revise based on comment

1. Support "1" as a boolean true value of the logRequests;
2. Treat old values of logRequests=false and logRequests=0
   as LogRequestsNever.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@afishbeck please review.

HPCCSmoketest commented:
Process of PR-12611, label: h20914 starts now.
The reason of this test is: New commit.
Commit ID: 33eff2e9393db0075c920cdb4c8d7df25fed3aa6
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 33eff2e9393db0075c920cdb4c8d7df25fed3aa6
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |23 sec  (00:00:23) |
|test (hthor) |835 |835 |0 |174 sec  (00:02:54) |
|test (thor) |759 |759 |0 |605 sec  (00:10:05) |
|test (roxie) |911 |911 |0 |238 sec  (00:03:58) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  213 sec (00:03:33) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1350 sec (00:22:30) |   20 sec (00:00:20) | 1635 sec (00:27:15) |

wangkx commented:
@afishbeck please review the last 2 commits.

HPCCSmoketest commented:
Process of PR-12611, label: h20914 starts now.
The reason of this test is: New commit.
Commit ID: d0de0082163cbafba37686f4d4265cdbf3b9b976
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: d0de0082163cbafba37686f4d4265cdbf3b9b976
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |835 |835 |0 |166 sec  (00:02:46) |
|test (thor) |759 |759 |0 |623 sec  (00:10:23) |
|test (roxie) |911 |911 |0 |234 sec  (00:03:54) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   36 sec (00:00:36) |  215 sec (00:03:35) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1354 sec (00:22:34) |   19 sec (00:00:19) | 1645 sec (00:27:25) |

wangkx commented:
@afishbeck please review.

HPCCSmoketest commented:
Process of PR-12611, label: h20914 starts now.
The reason of this test is: New commit.
Commit ID: cda8b0557d04da6702877b8dcfc108c48308c715
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: cda8b0557d04da6702877b8dcfc108c48308c715
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |835 |835 |0 |157 sec  (00:02:37) |
|test (thor) |759 |759 |0 |611 sec  (00:10:11) |
|test (roxie) |911 |911 |0 |229 sec  (00:03:49) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  211 sec (00:03:31) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1307 sec (00:21:47) |   19 sec (00:00:19) | 1588 sec (00:26:28) |

afishbeck reviewed (approved):
@wangkx looks good.
